### ISSUE TITLE
Spawn point labels have incorrect time span

### ISSUE THREAD

**Issue by dripdropper** [RocketMap/RocketMap#1944]
When you enable the display of spawn points under Marker Settings and hover over a spawn point, the times displayed are at least partially incorrect.  (Note that I'm assuming the use of speed scan after it has learned all of the spawn times.)

## Expected Behavior
The label for a spawn point should indicate the correct appearance and disappearance times for the spawn point.
In addition, the text stating "May appear as early as XX:XX" is confusing and should be removed or corrected (if it has any meaning in the current spawn environment).

## Current Behavior
An example label for a 30-minute spawn that is available from 21:03-51:03 is as follows:

>Every hour from 36:03 to 51:03
>May appear as early as 06:03

Here, the appearance time (36:03) is incorrect, the disappearance time (51:03) is correct, and the "may appear as early as" time makes no sense.

Every spawn point label seems to have a 15-minute time duration, which makes no sense in the current spawn environment where every(?) spawn is either 30 or 60 minutes.

## Possible Solution
It looks like the spawnpointLabel() function in map.js is hard-coded to display a 15-minute spawn duration based on an offset from a single input time value.  The function should be modified to receive two inputs and display the correct appearance and disappearance times.  Depending on the database structure, this could be the either the appearance and disappearance times directly, or just one of them and a duration.

I don't know the purpose of the "May appear as early as" text.  Maybe this was supposed to be used when a Pokemon had discontinuous available times (such as being shown for 15 minutes, hidden for 15 minutes, and then shown for 15 minutes)?  I'm not sure if discontinuous spawns exist any more (everything seems to be 30 or 60 minutes).  If they don't, this text can be removed.  If they do, this text should be reworded to be clearly specify what it means.

## Steps to Reproduce
1. Run speed scan until all the timers are learned
2. Go to Options -> Marker Settings and enable Spawn Points
3.  Place your mouse cursor over a spawn point

## Your Environment
* Version used:  Latest develop branch
* Environment name and version:  Python 2.7.12, MySQL/MariaDB 10.1
* Operating System and version:  Windows 7

**Comment by lluisd** [comment]
Yes, that code is outdated (the 15 min hardcoded times). I think we have the data (speed scan) but we don't put it in the right way to be displayed in the spawnpoints markers on the map

**Comment by EagleWorks** [comment]
Correct.  The source of the data is eventually from app.py at `if request.args.get('spawnpoints', 'false') == 'true':`.  There the `timestamp` value is passed with its value being `datetime.utcnow()`.  I haven't a clue where the correct times are stored, but if they could be passed there, the UI could be updated pretty easily.

**Comment by myntath** [comment]
Yes in models.py the get_spawnpoints is not taking advantage of speedscan at all, it could be a lot faster if the proper tables are queried. If you run a speedscan the data is stored in the spawnpoint table and should be readily accessible. Currently we are just checking every pokemon that has previously spawned and calculating the spawnpoints from that. I will have a look into this if I have time, but first will need to look into if we store spawnpoint data for the other scan types making querying previous spawns the only option for them.

**Comment by myntath** [comment]
I'm working on a fix which should work for now. I have tested it locally and performance is massively improved. The wording is either 'every hour from x to y' (no longer hardcoded but varies with length). Or times not finalised, current guess is x to y. 

However, I may have more to do before it's ready to create a pull request. I believe there are some very exotic spawn points which I haven't yet accounted for.

**Comment by dripdropper** [comment]
Thanks for working on this!

What sort of exotic spawn points do you think there might be?  I can try to help craft a reasonable message if you tell me what you're thinking.

Also, are we **sure** there are currently any spawn points other than continuous 30 or 60 minutes points?  I wouldn't be surprised if other detected spawn types weren't actually just bugs in the current speed scan algorithm (for example, see pull #1900).

**Comment by myntath** [comment]
I believe there are spawn points something like this shsh, where the same spawn appears for two 15 minute periods in the hour. Unfortunately I haven't mapped any of these to test. I think there are other spawns that also disappear and reappear within the hour as the same spawn. I think I can come up with the message, it's more a matter of sending through the right data and I don't have any test cases to help me.

A search of my database of 3134 spawns gives this
```SQL
select kind,links, count(*) from spawnpoint group by kind, links;

+------+-------+----------+
| kind | links | count(*) |
+------+-------+----------+
| hhhs | ????  |      234 |
| hhhs | hh??  |      190 |
| hhhs | hhh?  |        8 |
| hhss | hh??  |     2375 |
| hsss | h???  |        1 |
| ssss | +++-  |      326 |
+------+-------+----------+

```

This shows at least 15 minute and 45 spawns as well. They could be bugs but I think it is well established there is other spawn types.

Edit:
I have just cleared all spawn data and will now run a new initial scan using 1900. Will be interesting data to have anyway.

**Comment by dripdropper** [comment]
Here are the results from my database (without using pull #1900):
```
select kind,links, count(*) from spawnpoint group by kind, links;
+------+-------+----------+
| kind | links | count(*) |
+------+-------+----------+
| hhhs | ????  |      239 |
| hhhs | hh??  |       24 |
| hhhs | hhh?  |        3 |
| hhss | hh??  |     2138 |
| ssss | +++-  |       90 |
+------+-------+----------+
```
I have yet to witness anything other than 30 or 60 minute spawns after numerous days of usage, despite the results above suggesting I should see 15-minute spawns over 10% of the time.  I'm not saying they don't exist, but I have yet to see any evidence.  I'm curious to see what pull #1900 does.

**Comment by sirlindqvist** [comment]
I see 15 mins spawns sometimes, dunno if they are wrong? :)

**Comment by sebastienvercammen** [comment]
@myntath Don't link forks and ask people to use them. You should know our process by now. Prepare a PR that can gather feedback instead, and add "fixes #1944" in the description.

**Comment by myntath** [comment]
@sebastienvercammen Sorry seb, I will hold back next time until i'm ready to make a PR. 

@dripdropper I think you are right about the spawns. I only have four left to resolve and all the weird spawns have been disappearing with only one left. The other ~3000 are either 30/60 minute.
